pywemo/ouimeaux_device/api/service.py

```diff
@@ -29,73 +30,210 @@ class ActionException(Exception):
+        exception.
+
+        Args:
+            method: HTTP method/verb to use for request (ex: 'GET' or 'POST').
+            url: URL to to connect to.
+            retries: Number of retries, or urllib3.Retry instance.
+            timeout: Timeout in seconds for each request attempt.
+            kwargs: Additional arguments for urllib3 pool.request(**kwargs).
+
+        Raises:
+            requests.RequestException for any urllib3 exception or if the
+            response code is not 200.
+        """
+        if retries is None:
+            retries = self.retries
+        if timeout is None:
+            timeout = self.timeout
+
+        # Create and destroy the pool each time. WeMo devices do not support
+        # http keep-alive. Forcing the pool to be destroyed ensures that the
+        # http connection is also closed. This avoids tying up TCP sessions
+        # on the device.
+        with urllib3.PoolManager(retries=retries, timeout=timeout) as pool:
+            try:
+                response = pool.request(method=method, url=url, **kwargs)
+                if response.status != 200:
+                    raise urllib3.exceptions.HTTPError(
+                        f"Received status {response.status} for {url}"
+                    )
+            except urllib3.exceptions.HTTPError as err:
+                raise requests.RequestException(err)
+            return response
+
+    def get(self, url: str, **kwargs) -> urllib3.HTTPResponse:
+        """HTTP GET request."""
+        return self.request('GET', url, **kwargs)
+
+    def post(self, url: str, **kwargs) -> urllib3.HTTPResponse:
+        """HTTP POST request."""
+        return self.request('POST', url, **kwargs)
+
+    def urljoin(self, path: str) -> str:
+        """Build an absolute URL from a path."""
+        return urljoin(self._url, path)
+
+    @property
+    def url(self) -> str:
+        """Return the current URL for the session."""
+        return self._url
+
+    @url.setter
+    def url(self, url: str) -> str:
+        """Update the URL for the session."""
+        parsed_url = urlparse(url)
+        self._url = parsed_url.geturl()
+        self._port = parsed_url.port or 80
+        self._host = parsed_url.hostname
+        return url
+
+    @property
+    def port(self) -> int:
+        """TCP port associated with this session."""
+        return self._port
+
+    @property
+    def host(self) -> str:
+        """Hostname associated with this session."""
+        return self._host
+
+
 class Action:
     """Representation of an Action for a WeMo device."""
 
-    def __init__(self, device, service, action_config):
+    # A few actions take longer than the default timeout. Override the default
+    # timeout value for those actions.
+    soap_action_timeout_override = {
+        "urn:Belkin:service:bridge:1#AddDevice": 30,
+        "urn:Belkin:service:bridge:1#OpenNetwork": 30,
+        "urn:Belkin:service:WiFiSetup:1#GetApList": 10,
```

Comment:
10 is the default, is this one supposed to be 30 too?

Reply:
Nevermind, this is really for the retry so the default is 3 from the `Session` class.

Reply:
I changed the default for SOAP action requests to 3 seconds (`Session.timeout`). I've kept REQUESTS_TIMEOUT at 10 seconds as not all of the code that uses the requests library has retries enabled. In my tests `urn:Belkin:service:WiFiSetup:1#GetApList` completes in just over 4 seconds.